### PR TITLE
Relax diamond restriction on iarange nesting for enumeration

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #1391 

This partially relaxes the no-loops restriction on `iarange` nesting during enumeration. After this PR, there is still a restriction involving `iarange` nesting, and I believe this is optimal for the complexity class. I am working on a parsimonious statement of this restriction in #1391 .

The change in this PR was basically to compute the `parent` ordinal independently for each connected component of tensors at each ordinal, rather than compute a single `parent` for each ordinal.

## Tested

- [x] refactoring is covered by existing tests
- [x] added a regression tests for `ubersum`
- [x] added a regression test for `TraceEnum_ELBO`

### CODE REVIEW

File: pyro/infer/contract.py [hunk 1]
```diff
@@ -213,52 +213,6 @@ def broadcast(self, term, ordinal):
         return term
 
 
-def _check_tree_structure(tensor_tree):
-    # Check for absence of diamonds, i.e. collider iaranges.
-    # This is required to avoid loops in message passing.
-    #
-    # For example let f1, f2 be two CondIndepStackFrames, and consider
```
Comment (fritzo) [inline pyro/infer/contract.py]: This is a nice figure but is no longer correct. I'll try to add something like it to both the tutorial #1355 and the tech report #1391 .

File: tests/infer/test_enum.py [hunk 2]
```diff
@@ -2328,12 +2328,141 @@ def guide(data):
         elbo.differentiable_loss(model_iarange_iarange, guide, data)
 
 
+@pytest.mark.parametrize('scale', [1, 10])
+def test_elbo_enumerate_iaranges_7(scale):
```
Comment (fritzo) [inline tests/infer/test_enum.py]: The following example motivates this PR.

**Comment by neerajprad** [comment]
Looks great! I checked that this works with one of the dag examples in `mcmc.test_valid_models`, as well. I would like to dig in a bit more, and understand how the `Expected tree-structured iarange nesting..` warning is triggered with this refactor in another pass shortly.

**Comment by fritzo** [comment]
@neerajprad let me first add a test for `ubersum("ae,be,abe->", ..., batch_dims="ab")` and make sure we can handle this.

**Pull request merged**